dex/elrond_dex_router/src/pair_manager.rs

```diff
@@ -0,0 +1,206 @@
+use elrond_dex_pair::fee::ProxyTrait as _;
+
+#[elrond_wasm::module]
+pub trait PairManagerModule:
+    state::StateModule + factory::FactoryModule + token_send::TokenSendModule
+{
+    #[only_owner]
+    #[endpoint(setFeeOn)]
+    fn set_fee_on(
+        &self,
+        pair_address: Address,
+        fee_to_address: Address,
+        fee_token: TokenIdentifier,
+    ) -> SCResult<()> {
+        require!(self.is_active(), "Not active");
+        self.check_is_pair_sc(&pair_address)?;
+
+        self.pair_contract_proxy(pair_address)
+            .set_fee_on(true, fee_to_address, fee_token)
+            .execute_on_dest_context();
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[endpoint(setFeeOff)]
+    fn set_fee_off(
+        &self,
+        pair_address: Address,
+        fee_to_address: Address,
+        fee_token: TokenIdentifier,
+    ) -> SCResult<()> {
+        require!(self.is_active(), "Not active");
+        self.check_is_pair_sc(&pair_address)?;
+
+        self.pair_contract_proxy(pair_address)
+            .set_fee_on(false, fee_to_address, fee_token)
+            .execute_on_dest_context();
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint(acceptPay)]
+    fn accept_pay(&self) {}
+
+    #[payable("*")]
+    #[endpoint(multiPairSwap)]
+    fn multi_pair_swap(
+        &self,
+        #[payment_token] token_id: TokenIdentifier,
+        #[payment_amount] amount: Self::BigUint,
+        #[payment_nonce] nonce: Nonce,
+        swap_operations: MultiArgVec<SwapOperationType<Self::BigUint>>,
+        #[var_args] opt_accept_funds_func: OptionalArg<BoxedBytes>,
+    ) -> SCResult<()> {
+        require!(nonce == 0, "Invalid nonce. Should be zero");
+        require!(amount > 0, "Invalid amount. Should not be zero");
+        require!(
+            !swap_operations.is_empty(),
+            "Invalid swap operations chain. Should not be empty"
+        );
+
+        let caller = self.blockchain().get_caller();
+        let mut residuum_vec = Vec::new();
+        let mut last_received_token_id = token_id;
+        let mut last_received_amount = amount;
+
+        for entry in swap_operations.into_vec() {
+            let (pair_address, function, token_wanted, amount_wanted) = entry.into_tuple();
+            self.check_is_pair_sc(&pair_address)?;
+
+            if function == BoxedBytes::from(SWAP_TOKENS_FIXED_INPUT_FUNC_NAME) {
+                let token_amount_out = self.actual_swap_fixed_input(
+                    pair_address,
+                    last_received_token_id,
+                    last_received_amount,
+                    token_wanted,
+                    amount_wanted,
+                );
+                last_received_token_id = token_amount_out.token_id;
+                last_received_amount = token_amount_out.amount;
+            } else if function == BoxedBytes::from(SWAP_TOKENS_FIXED_OUTPUT_FUNC_NAME) {
+                let (token_amount_out, residuum) = self.actual_swap_fixed_output(
```

Comment:
The problem that I foresee here is that a user would end up with some dust from a token he doesn't want. Although it would be quite hard to do, we might need to think of a solution to have this dust at least in the first input token. Does UniswapV2 handles this residuum?

Reply:
I don't see a problem here. Having extra tokens is not something bad. Let's not overcomplicate. If he strictly does not want it, he can only swapTokensFixedInput and that's it. Uniswap does not handle any residuum I think, although it could be double checked.